### ISSUE TITLE
Getting OCIError: ORA-24920: column size too large for client when accessing a table in Oracle 12c with 32k Varchar2 columns

### ISSUE THREAD

**Issue by dsalter** [kubo/ruby-oci8#82]
So kind of new to the ruby environment, but I was migrating a ruby on rails application from a Oracle 11i database to a Oracle 12c database and at the same time modified some varchar2(4k) columns to the new 12c max varchar2(32K).  I was able to do the migration, but when I went to access the fields in ruby on rails I was getting the following message in an error log:

OCIError: ORA-24920: column size too large for client.

We are currently using the ruby-oci8 gem version 2.1.0 and I am getting ready to test the 2.1.8 version, but thought I would open an issue to see if anyone else has hit this.  I have searched the internet and can't find anything around this.  Since this is the first issue that I have created on this site, please let me know if there is any more information you need around this issue.  Any information around this would be appreciated.

**Comment by kubo** [comment]
In https://docs.oracle.com/database/121/ERRMG/e24280.htm#ORA-24920
> Action: Upgrade the client stack to Oracle 12c or higher.

**Comment by dsalter** [comment]
Hey Kubo,

I am using the Oracle 12c stack.  I did however download the 2.1.8 version of ruby-oci8 and with just a little testing that seems to have fixed the issue.  I will be doing more testing today and will post an update once I have completed that.  Just an FYI, I was also getting the error in the 2.1.5 version as well so looks like between 2.1.5 and 2.1.8 the issue might have been resolved.

**Comment by kubo** [comment]
Hi,

Okay, I'll check it later, probably this weekend.
Well, just to be sure, could you check `OCI8::oracle_client_version` for each ruby-oci8 installation?

**Comment by dsalter** [comment]
Hey Kubo,

So I did more testing today with both ruby-oci8 2.1.5 and 2.1.8.  I am sure now that 2.1.8 has corrected the issue and continued to get the error with 2.1.5.  Since 2.1.8 is the most current version I guess I can close this issue.  One note and not a big deal, but the ruby-oci8 2.1.8 gem does not include 12c in the description of version of Oracle it supports.  I just wanted to thank you for all your comments on this issue and help you provided.  Thanks again.

**Comment by kubo** [comment]
After I installed Oracle 12.1.0.2 and set `max_string_size=EXTENDED`,
I tested the following script with ruby-oci8 2.1.0, 2.1.5 and 2.1.8 and got success for all versions.
```
require 'oci8'

conn = OCI8.new('ruby/oci8')
conn.exec('drop table test_table purge') rescue nil
conn.exec('create table test_table (col varchar2(32767))')
conn.exec('insert into test_table values(:1)', '1' * 32767)

puts "ruby-oci8 version: #{OCI8::VERSION}"
puts "oracle client version: #{OCI8::oracle_client_version}"
puts "oracle server version: #{conn.oracle_server_version}"
if conn.select_one('select * from test_table')[0].length == 32767
  puts "success"
else
  puts "error"
end
```
one of results:
```
ruby-oci8 version: 2.1.0
oracle client version: 12.1.0.2.0
oracle server version: 12.1.0.2.0
success
```
@dsalter
Could you run the script with your ruby-oci8 2.1.0 and 2.1.5 if you can?

**Comment by dsalter** [comment]
Hey Kubo,

I will see if I can get to that this week and let you know.
  Thanks,
          Daryl

**Comment by dsalter** [comment]
Hey Kubo,

So I got the same results as you did:

ruby-oci8 version: 2.1.0
oracle client version: 12.1.0.2.0
oracle server version: 12.1.0.2.0
success

ruby-oci8 version: 2.1.5
oracle client version: 12.1.0.2.0
oracle server version: 12.1.0.2.0
success

ruby-oci8 version: 2.1.8
oracle client version: 12.1.0.2.0
oracle server version: 12.1.0.2.0
success

But that is not what I am seeing when I run our app though rails.  I am going to try and create a little rails app that does what your script does and see if I can get the error.  So the only thing I am changing is the ruby-oci8 module when I am seeing the error in 2.1.5 vs not seeing it with 2.1.8.  I will send an update once I can try and test it though rails.

  Daryl

**Comment by dsalter** [comment]
Ok this is weird, so in my dev environment when I installed 2.1.8 I uninstalled 2.1.0 and 2.1.5, to do the testing above I reinstalled 2.1.0 and 2.1.5 and now the app works with both of those (but they did not work before I uninstalled them).  In my staging environment I installed 2.1.8 but did not uninstall 2.1.5, in that environment the app works with 2.1.8 but does not work with 2.1.5.  So not sure what is going on, looks like if I uninstall the old versions and reinstall them they start to work (like I did in the dev environment) but if I do not uninstall them only 2.1.8 seems to work (like my staging environment.  I will continue to research this and let you know what I find out.

  Daryl

**Comment by dsalter** [comment]
So I think we (a ruby developer and I) might have figured it out.  Here is what we think is happening, the reason the 2.1.5 started working in my dev environment is due to the fact that I uninstalled the 2.1.5 (that was installed when we were on 11g) and reinstalled 2.1.5 (which now is using 12c).  I think somehow the old install was pointing to the 11g oci even though we said our ORACLE_HOME was 12c and the new install of 2.1.5 pointed to the new 12c oci which is why it started working.  Going to run one more test tomorrow to confirm this, but it looks like you need to reinstall the ruby-oci8 gem when switching between 11g and 12c.  So in our staging environment I did not uninstall the 2.1.5 gem when I installed the 2.1.8 gem and pointing to 2.1.8 the app works fine but pointing to the old 2.1.5 (installed when we were using 11g) I get this error when I run your script:

stmt.c:230:in oci8lib_191.so: ORA-01461: can bind a LONG value only for insert into a LONG column (OCIError)
        from /home/syt/jlewis/.rvm/gems/ruby-1.9.3-p125@rails322/gems/ruby-oci8-2.1.5/lib/oci8/cursor.rb:129:in `exec'
        from /home/syt/jlewis/.rvm/gems/ruby-1.9.3-p125@rails322/gems/ruby-oci8-2.1.5/lib/oci8/oci8.rb:278:in `exec_internal'
        from /home/syt/jlewis/.rvm/gems/ruby-1.9.3-p125@rails322/gems/ruby-oci8-2.1.5/lib/oci8/oci8.rb:269:in `exec'

If you have any thoughts on this please let me know.

Thanks again,
Daryl

**Comment by kubo** [comment]
Thanks for the investigation.

> somehow the old install was pointing to the 11g oci even though we said our ORACLE_HOME was 12c and the new install of 2.1.5 pointed to the new 12c oci which is why it started working.

The location of OCI libraries is embedded in oci8lib_???.so on installation in Linux and Solaris.
The search order of library load path in Linux is:

1. RPATH embedded in oci8lib_???.so.

   ```shell
   $ objdump -p ext/oci8/oci8lib_200.so | grep RPATH
   RPATH                /home/kubo/local/ruby64-2.0.0-p0/lib:/home/ora112/app/product/11.2.0/dbhome_1/lib
   ```
2. the environment variable LD_LIBRARY_PATH
3. /etc/ld.so.conf

Even though ORACLE_HOME and LD_LIBRARY_PATH are set to new location, old install points to the old location.

The following script will print 11g for old 2.1.5 (that was installed when you were on 11g).
```ruby
require 'oci8'
puts "ruby-oci8 version: #{OCI8::VERSION}"
puts "oracle client version: #{OCI8::oracle_client_version}"
```

**Comment by dsalter** [comment]
Hey Kubo,

So thanks for all the help and I ran you script and it came back with 11g when I ran against a gem that was installed when we only had 11g.  It could be that I am new to ruby on rails, but it might be nice if in the documentation it noted that if you are installing a new version of oracle and using a new feature that the gem ruby-oci8 would need to be reinstalled while you are pointing to the new oracle binaries.  

Thanks again for all the support.
        Daryl
